### PR TITLE
QPACK and naive H3 implementation

### PR DESCRIPTION
**Pull request by stammw**
Continuing the work of @evolix1, from #20, which is blocking #130.

Features:
* [x] Encoder
* [x] Reference tracking
* [x] Track DynamicTable::fields mutation, update DynamicTable::*_map when constructing DynamicTableEncoder instead of rebuilding HashMaps
* [-] N bit handling
* [-] Encode / decode concurreny (needs to have concurrent shared mutable state...)
* [ ] Configuration (may need H/3 impl to start)

Optimization: 
* [ ] Improve Literal argument passing: 
   * ```LiteralWithPostBaseNameRef::new(index, field.value.clone())```: No
   * Look for a solution in https://deterministic.space/elegant-apis-in-rust.html, https://phaazon.net/blog/on-owning-borrowing-pub-interface
* [ ] Header fields lifetimes for Cow
* [ ] Use Reference count for DynamicTable::fields DynamicTable::*_map

### CODE REVIEW

File: quinn-h3/src/qpack/block.rs [hunk 1]
```diff
@@ -0,0 +1,365 @@
+use bytes::{Buf, BufMut};
+
+use super::prefix_int;
+use super::prefix_string;
+use super::ParseError;
+
+pub(super) enum HeaderBlocField {
```
Comment (Ralith) [inline quinn-h3/src/qpack/block.rs]:
```suggestion
pub(super) enum HeaderBlockField {
```

Comment (Ralith) [inline quinn-h3/src/qpack/block.rs]: Is it really necessary to use `pub(super)` in this module, given that the module is not itself reexported?

File: quinn-h3/src/qpack/decoder.rs [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+    Duplicate, DynamicTableSizeUpdate, EncoderInstruction, HeaderAck, InsertCountIncrement,
+    InsertWithNameRef, InsertWithoutNameRef, StreamCancel,
+};
+use super::ParseError;
+
+use super::prefix_int;
+use super::prefix_string;
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
```
Comment (Ralith) [inline quinn-h3/src/qpack/decoder.rs]: If this error (and the others like it) ever reach a public API, it'd be nice to derive `Display` and `Error` impls for them using err-derive (see e.g. `quinn_proto::ConnectionError` for an example)

File: quinn-h3/src/qpack/vas.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
+/**
+ * https://tools.ietf.org/html/draft-ietf-quic-qpack-01#section-2.2.1
+ * https://tools.ietf.org/html/draft-ietf-quic-qpack-01#section-2.2.2
+ */
+
+/*
+ *  # Virtualy infinit address space mapper.
```
Comment (Ralith) [inline quinn-h3/src/qpack/vas.rs]:
```suggestion
 *  # Virtually infinite address space mapper.
```

File: quinn-h3/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "quinn-h3"
+version = "0.0.1"
+authors = ["Benjamin Saunders <ben.e.saunders@gmail.com>", "Dirkjan Ochtman <dirkjan@ochtman.nl>"]
```
Comment (djc) [inline quinn-h3/Cargo.toml]: You should add yourself here!

**Comment by stammw** [comment]
@djc I did not think of that. I'm not sure the common code surface would be that big. Maybe for the primitive types like `prefix_string`.

**Comment by djc** [comment]
I think that's what I was looking at, but I understand if there's not enough overlap. 👍

**Comment by stammw** [comment]
@djc Regarding the allow(dead_code) usages, do you prefer that I delete them and leave the build yield warnings until the functions are used by the H3 impl?

**Comment by stammw** [comment]
I think we're ready to merge ! 

The interop tool has been updated with only a very simple H3 scenario (with no encoder stream, nor receiving anything from server other than the response). 

Trying to implement a full-featured client, I started to think it was too much code to write in form of a quick implementation. So it would be better to continue writing a solid code into another PR.

File: interop/src/main.rs [hunk 5]
```diff
@@ -176,6 +182,50 @@ fn run(log: Logger, options: Opt) -> Result<()> {
...
+            .and_then(|conn| {
+                assert!(state.lock().unwrap().saw_cert);
+                handshake = true;
+                let conn = conn.connection;
+                let control_stream = conn.open_uni();
+                let control_fut = control_stream
+                    .map_err(|e| format_err!("failed to open control stream: {}", e))
+                    .and_then(move |stream| {
+                        let mut buf = BytesMut::new();
+                        buf.put(0x43u8);
```
Comment (djc) [inline interop/src/main.rs]: What is this for?

**Comment by Ralith** [comment]
I'll try to have a look at the futuresy bits tomorrow per @stammw's request, at least.

File: interop/src/main.rs (hunk 5 above)
Comment (stammw) [reply interop/src/main.rs]: The control stream is useful to send one's endpoint settings https://quicwg.org/base-drafts/draft-ietf-quic-http.html#rfc.section.3.2.1. This is not mandatory for this code to work but I thought I would leave it as is to validate that the `SETTINGS FRAME` is correctly encoded.

Normally it shouldn't be chained with the request, but executed in parallel. This is something I plan to implement in future HTTP lifecycle module. Moreover the implementation should receive one `SETTINGS FRAME`  from the server and change it's QPACK encoder behavior accordingly. Which will involve shared mutable state.

Comment (Ralith) [reply interop/src/main.rs]: Does it need to involve a magic number hardcoded into the interop tool?

File: interop/src/main.rs [hunk 6]
```diff
@@ -69,7 +72,10 @@ fn run(log: Logger, options: Opt) -> Result<()> {
     tls_config
         .dangerous()
         .set_certificate_verifier(Arc::new(InteropVerifier(state.clone())));
-    tls_config.alpn_protocols = vec![str::from_utf8(quinn::ALPN_QUIC_HTTP).unwrap().into()];
+    tls_config.alpn_protocols = vec![
+        str::from_utf8(quinn::ALPN_QUIC_HTTP).unwrap().into(),
+        str::from_utf8(quinn::ALPN_QUIC_H3).unwrap().into(),
+    ];
```
Comment (Ralith) [inline interop/src/main.rs]: We should have two separate client configs, one for `hq-whatever` and one for `h3-whatever`. Placing both in a single config doesn't mean you can use either, it means you're willing to let the server select either for you, and it's your responsibility to comply with the server's selection. Plus, this way you can enable/disable incoming unidirectional streams for each as appropriate.

Also, we can get rid of the `str::from_utf8(...).unwrap()` now.

File: interop/src/main.rs [hunk 7]
```diff
@@ -176,6 +182,50 @@ fn run(log: Logger, options: Opt) -> Result<()> {
...
+                    .and_then(|req_stream| h3_get(req_stream))
+                    .and_then(move |data| {
+                        println!(
+                            "read {} bytes: \n\n{}\n\n closing",
+                            data.len(),
+                            String::from_utf8_lossy(&data)
+                        );
+                        conn.close(0, b"done").map_err(|_| unreachable!())
+                    });
+                control_fut.and_then(|_| req_fut).map(|_| h3 = true)
```
Comment (Ralith) [inline interop/src/main.rs]: You can make these concurrent by replacing `and_then` with `join`

File: interop/src/main.rs [hunk 8]
```diff
@@ -197,12 +247,74 @@ fn run(log: Logger, options: Opt) -> Result<()> {
...
+    match HttpFrame::decode(&mut cur) {
+        Ok(HttpFrame::Headers(text)) => {
+            let mut resp_block = std::io::Cursor::new(&text.encoded);
+            match qpack::decode_header(table, &mut resp_block) {
+                Ok(_) => (),
+                Err(e) => return Err(format_err!("failed to decode response header {}", e)),
+            }
+        }
+        Ok(f) => {
+            println!("response basd frame type {:?}", f);
```
Comment (Ralith) [inline interop/src/main.rs]:
```suggestion
            println!("response bad frame type {:?}", f);
```

File: interop/src/main.rs [hunk 9]
```diff
@@ -176,6 +182,50 @@ fn run(log: Logger, options: Opt) -> Result<()> {
...
     }
 
+    let mut h3 = false;
+    let result = runtime.block_on(
+        endpoint
+            .connect_with(&client_config, &remote, &options.host)?
+            .map_err(|e| format_err!("failed to connect: {}", e))
+            .and_then(|conn| {
+                assert!(state.lock().unwrap().saw_cert);
+                handshake = true;
```
Comment (Ralith) [inline interop/src/main.rs]: Copy-paste error? These two lines are redundant to their prior instance and should be removed here.

File: interop/src/main.rs [hunk 10]
```diff
@@ -176,6 +182,50 @@ fn run(log: Logger, options: Opt) -> Result<()> {
         }
     }
 
+    let mut h3 = false;
+    let result = runtime.block_on(
+        endpoint
+            .connect_with(&client_config, &remote, &options.host)?
```
Comment (Ralith) [inline interop/src/main.rs]:
```suggestion
            .connect_with(&client_config, &remote, host)?
```
Looks like you missed this in a rebase.

File: quinn-proto/src/lib.rs [hunk 11]
```diff
@@ -54,6 +54,9 @@ pub const VERSION: u32 = 0xff00_0011;
 /// TLS ALPN value for HTTP over QUIC
 pub const ALPN_QUIC_HTTP: &[u8] = b"hq-17";
 
+/// TLS ALPN value for H3
+pub const ALPN_QUIC_H3: &[u8] = b"h3-17";
```
Comment (Ralith) [inline quinn-proto/src/lib.rs]:
```suggestion
pub const ALPN_QUIC_H3: &[u8] = b"h3-18";
```
As of #225 (whoops, looks like I forgot this above too, feel free to get them both)

File: interop/src/main.rs (hunk 5 above)
Comment (Ralith) [reply interop/src/main.rs]: re: shared mutable state: it's acceptable to have `Arc<Mutex<State>>` (perhaps hidden in a helper type that the user constructs per-stream?) for cases like this, IMO.

Comment (djc) [reply interop/src/main.rs]: Yeah, I was also mostly responding to the magic number, which seems a bit awkward and hard to understand.

Comment (stammw) [reply interop/src/main.rs]: Ah yes, it's not very clean indeed...

Comment (stammw) [reply interop/src/main.rs]: fixed in 51521a0

File: interop/src/main.rs (hunk 7 above)
Comment (stammw) [reply interop/src/main.rs]: Thank you, it's actually serially executed on purpose. Not making any request if the control stream fails.

File: interop/src/main.rs (hunk 6 above)
Comment (stammw) [reply interop/src/main.rs]: fixed in 4ee65b9

File: interop/src/main.rs (hunk 10 above)
Comment (stammw) [reply interop/src/main.rs]: fixed in 4ee65b9

File: interop/src/main.rs (hunk 9 above)
Comment (stammw) [reply interop/src/main.rs]: fixed in 4ee65b9

File: quinn-proto/src/lib.rs (hunk 11 above)
Comment (stammw) [reply quinn-proto/src/lib.rs]: fixed in 8569932

**Comment by djc** [comment]
🎉

**Pull request merged**